src/env/tock/clock.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright 2022-2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::api::clock::Clock;
+use libtock_core::syscalls;
+
+mod command_nr {
+    pub const GET_CLOCK_FREQUENCY: usize = 1;
+    pub const GET_CLOCK_VALUE: usize = 2;
+}
+
+const DRIVER_NUMBER: usize = 0x00000;
+
+pub struct TockTimer {
+    end_epoch: usize,
+    end_tick: usize,
+}
+
+/// Returns a tupel of u24 sum and how many times the sum wraps.
+fn add_to_u24_with_wraps(lhs: usize, rhs: usize) -> (usize, usize) {
+    // Saturating should never happen, but it fails gracefully.
+    let sum = lhs.saturating_add(rhs);
+    (sum & 0xffffff, sum >> 24)
+}
+
+fn wrapping_sub_u24(lhs: usize, rhs: usize) -> usize {
+    lhs.wrapping_sub(rhs) & 0xffffff
+}
+
+pub struct TockClock {
+    epoch: usize,
```

Comment:
I am not convinced by this 32-bits epoch design. I would either use no epoch (i.e. should tickle more often than 0x800000 ticks) or store epoch in high 8 bits and wrap around providing 32-bits clock (i.e. should tickle more often than 0x80000000 ticks).

If I'm not mistaken, the Tock clock is 32kHz and 24-bits. So if we want to tickle every time bit 23 flips, this gives 256 seconds. I think we can safely assume that we tickle that frequently. So I would not use any epoch.

Note also that we can use the full 24-bits if we want (going to 512 seconds), by storing an additional reference point. But I think that's waste of memory.

Reply:
Actually, thinking about it, we anyway need to tickle every 256 seconds for the epoch to be correctly updated. So the only benefit of epoch seems to be to support more than 24-bits clock and thus more than 23-bits duration. Is there anything in CTAP that requires more than 256 seconds duration?

Reply:
We don't need longer timeout, it was just a nice-to-have. The 32-bit design sounds like it should be as good, so I'll look into it.

Reply:
I realized that 32 bit timers have a problem: While `tickle` gives us a correct clock, the timers only get tickled when checked explicitly. No problem before, but if you i.e. create a `u32` auth token timer, and then don't talk to the authenticator for a full wrap, the timer doesn't notice.

The clock could fix this by keeping track of all timers it created. Then the `Timer` would be a `u16` identifier and the `Clock` has a FIFO queue of timer identifiers, together with their `u32` end tick. But that's probably not better or easier than just using u64 timers.

(The latest PR doesn't change the Timer design, just improves the implementation a bit.)

Reply:
> The clock could fix this by keeping track of all timers it created.

Just to be sure I follow, the idea would be to stop using `Ctap::update_timeouts` in the main loop? This is currently where timers are being kept track of (and thus regularly tickled).

If yes, then indeed we would need at least 48-bits timers to cover 100 years (and be on the safe side regarding wrapping).

If no, then 24-bits timers should be enough if we tickle every 4 minutes (32-bits if we want 18 hours).

Reply:
`Ctap::update_timeouts` would need to periodically check all timers that exist in your code. On of the ideas of the new design is that it is easier to use: Timer behavior can be understood locally, and don't need to be registered / tickled anywhere explicitly. Even before this PR, the timestamps in HID were not checked or tickled and had the wrapping failure mode IIRC. Because it is was easy to do timers wrong.

Reply:
Ok, so if the goal is to remove `Ctap::update_timeouts` and expect the environment to have absolute time, then I think the API should be simplified to be consistent with this decision:
- We don't need `Clock::check_timer`, `Clock::update_timer`, and `Clock::tickle`. We just need `Clock::is_elapsed`.
- We could even change `Clock::make_timer` into `Clock::now` and just use timestamps. Probably better.
- The `TockEnv` implementation (and not the `Env` abstraction) would expose a tickle function to be called from main, as is currently the case.

Reply:
> We only need timers

If we only need timers, then we need to keep track of timers. Because not keeping track of timers essentially means keeping track of absolute time, which means timestamp. This is why I say we should be consistent. The API is either about timestamps (and doesn't need to keep track of timers), or it is about timers (and needs to keep track of timers).

The keeping track of timers can either be done within `Ctap` or within `Env`. Currently it is done within `Ctap` with this `update_timeouts` function.

I don't really mind on the solution. Here are how they all compare:
1. Timestamps: Clean API. Easy to implement with 64-bits clock.
2. Timers in `Ctap`: Not clean API. Easy to implement because pretty close to current state.
3. Timers in `Env`: Clean API. Needs a little bit of implementation to keep track of timers.

Reply:
The difference between 2 and 3 for you is whether timers trigger a callback themselves on timeout?

Reply:
No, there's no callback. The difference is whether `Ctap` (the library) is responsible for keeping track of timers (and checking them regularly) or if `Env` is (thus checking them on tickle).

Reply:
For example, (3) would be closer to what Tock is doing. Tock keeps track of timers because the callback tells you which timer triggered AFAICT.

Reply:
We discussed in the resolved comment that we can remove `Ctap::update_timeouts` in a follow-up PR. Together with `is_elapsed`, wouldn't that be a clean API for timers?

Reply:
The Clock documentation now states that timer shall not wrap when left untouched.